llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp

```diff
@@ -1435,6 +1435,33 @@ static Instruction *foldBitOrderCrossLogicOp(Value *V,
   return nullptr;
 }
 
+static Value *simplifyReductionOperand(Value *Arg, bool CanReorderLanes) {
+  if (!CanReorderLanes)
+    return nullptr;
+
+  Value *V;
+  if (match(Arg, m_VecReverse(m_Value(V))))
+    return V;
+
+  ArrayRef<int> Mask;
+  if (!isa<FixedVectorType>(Arg->getType()) ||
+      !match(Arg, m_Shuffle(m_Value(V), m_Undef(), m_Mask(Mask))) ||
+      !cast<ShuffleVectorInst>(Arg)->isSingleSource())
+    return nullptr;
+
+  int Sz = Mask.size();
+  SmallBitVector UsedIndices(Sz);
+  for (int Idx : Mask) {
+    if (Idx == PoisonMaskElem || UsedIndices.test(Idx))
+      return nullptr;
+    UsedIndices.set(Idx);
+  }
+
+  // Can remove shuffle iff just shuffled elements, no repeats, undefs, or
+  // other changes.
+  return UsedIndices.all() ? V : nullptr;
```

Comment:
Think the shuffle check could be simplified with `getShuffleDemandedElts`.

Then the check would just be:
```
if(DemandedLHS.isAllOnes() && DemandedRHS.isZero()) {
  return LHS;
}
if(DemandedRHS.isAllOnes() && DemandedLHS.isZero()) {
  return RHS;
}
return nullptr;
```

Reply:
Hmm, this is not a NFC change. I am not quite sure how to use `getShuffleDemandedElts`, but I'm pretty sure your example above now permits use of a RHS whereas previously it didn't (see `!match(Arg, m_Shuffle(m_Value(V), m_Undef(), m_Mask(Mask)))`). I would prefer to limit this to refactoring only as it's not relevant to this patch. I can keep the `if(DemandedLHS.isAllOnes() ...` check.

Reply:
I had a look at using `getShuffleDemandedElts` and the code didn't immediately look much better as I still need to keep most of the existing checks, i.e. check for `isSingleSource` and so on. For now, I'd prefer to leave the code as it is, but I'm happy to revisit this in a follow-on patch. Just out of curiosity, have you seen any examples where we pass through the 2nd operand of the shufflevector? i.e. `reduce_or(shufflevector(<4 x i32> %a, <4 x i32> %b, <4 x i32> <7, 5, 6, 4>))`. I'd expect this to be canonicalised to `reduce_or(shufflevector(<4 x i32> %a, <4 x i32> undef, <4 x i32> <3, 1, 2, 0>))`